frame/conviction-voting/src/types.rs

```diff
@@ -0,0 +1,235 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Miscellaneous additional datatypes.
+
+use sp_std::marker::PhantomData;
+
+use super::*;
+use crate::{AccountVote, Conviction, Vote};
+use codec::{Codec, Decode, Encode};
+use frame_support::{
+	traits::VoteTally, CloneNoBound, DefaultNoBound, EqNoBound, PartialEqNoBound,
+	RuntimeDebugNoBound,
+};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{Saturating, Zero},
+	RuntimeDebug,
+};
+
+/// Info regarding an ongoing referendum.
+#[derive(
+	CloneNoBound,
+	DefaultNoBound,
+	PartialEqNoBound,
+	EqNoBound,
+	RuntimeDebugNoBound,
+	TypeInfo,
+	Encode,
+	Decode,
+)]
+#[scale_info(skip_type_params(Total))]
+pub struct Tally<
+	Votes: Clone + Default + PartialEq + Eq + sp_std::fmt::Debug + TypeInfo + Codec,
+	Total,
+> {
+	/// The number of aye votes, expressed in terms of post-conviction lock-vote.
+	pub ayes: Votes,
+	/// The number of nay votes, expressed in terms of post-conviction lock-vote.
+	pub nays: Votes,
+	/// The amount of funds currently expressing its opinion. Pre-conviction.
+	pub turnout: Votes,
+	/// Dummy.
+	dummy: PhantomData<Total>,
+}
+
+impl<
+		Votes: Clone
+			+ Default
+			+ PartialEq
+			+ Eq
+			+ sp_std::fmt::Debug
+			+ Copy
+			+ AtLeast32BitUnsigned
+			+ TypeInfo
+			+ Codec,
+		Total: Get<Votes>,
+	> VoteTally<Votes> for Tally<Votes, Total>
+{
+	fn ayes(&self) -> Votes {
+		self.ayes
+	}
+
+	fn turnout(&self) -> Perbill {
+		Perbill::from_rational(self.turnout, Total::get())
+	}
+
+	fn approval(&self) -> Perbill {
+		Perbill::from_rational(self.ayes, self.ayes.saturating_add(self.nays))
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn unanimity() -> Self {
+		Self { ayes: Total::get(), nays: Zero::zero(), turnout: Total::get(), dummy: PhantomData }
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn from_requirements(turnout: Perbill, approval: Perbill) -> Self {
+		let turnout = turnout.mul_ceil(Total::get());
+		let ayes = approval.mul_ceil(turnout);
+		Self { ayes, nays: turnout - ayes, turnout, dummy: PhantomData }
+	}
+}
+
+impl<
+		Votes: Clone
+			+ Default
+			+ PartialEq
+			+ Eq
+			+ sp_std::fmt::Debug
+			+ Copy
+			+ AtLeast32BitUnsigned
+			+ TypeInfo
+			+ Codec,
+		Total: Get<Votes>,
+	> Tally<Votes, Total>
+{
+	/// Create a new tally.
+	pub fn new(vote: Vote, balance: Votes) -> Self {
+		let Delegations { votes, capital } = vote.conviction.votes(balance);
+		Self {
+			ayes: if vote.aye { votes } else { Zero::zero() },
+			nays: if vote.aye { Zero::zero() } else { votes },
+			turnout: capital,
+			dummy: PhantomData,
+		}
+	}
+
+	pub fn from_parts(ayes: Votes, nays: Votes, turnout: Votes) -> Self {
+		Self { ayes, nays, turnout, dummy: PhantomData }
+	}
+
+	/// Add an account's vote into the tally.
+	pub fn add(&mut self, vote: AccountVote<Votes>) -> Option<()> {
```

Comment:
This should return a result, else the compiler will not warn the user when they do not use the returned value, and thus may not handle the error case well.